Measure V2

bramstroker opened the pull request:
Start with new cleaner architecture for measuring. DRY code and allow to switch power meters and light controllers easily without changing the measure logic

Todo:
- [x] Switchable light controllers / power meters using some selection procedure / factories.
- [x] Move sensitive configuration which should not be commited to .env file
- [x] Implement Kasa powermeter
- [x] Implement Tasmota powermeter
- [x] Allow user to select light entity_id from a list
- [x] Implement some error handling in the API responses
- [x] Retry mechanism in hass power meter to wait for a recent measurement
- [x] Take multiple measurement and pick the average (see for example current kasa implementation).
- [x] Automatic gzipping of CSV files
- [x] Ask measure information so we can add this in model.json
- [x] Format model.json
- [x] Validate power meter and light controller configurations 
- [x] Create requirements.txt
- [x] Compare timestamps to see if measurement is not outdated

stefangries commented:
Thank you for your work. A few more notes about the readme:

- On Windows you can't run "source measure/bin/activate".  Source does not exist here.
- The requirements.txt is still missing

Unfortunately, measure.py cannot be executed correctly on my system either.

![image](https://user-images.githubusercontent.com/10342151/131263693-f079ec86-d855-40bb-8bcf-fee6e212ddca.png)

bramstroker commented:
> Thank you for your work. A few more notes about the readme:
> 
> * On Windows you can't run "source measure/bin/activate".  Source does not exist here.

Thanks, will add a note that creating a virtual env is optional.
> * The requirements.txt is still missing
> 

Yes still need to add it.
> Unfortunately, measure.py cannot be executed correctly on my system either.
> 
> ![image](https://user-images.githubusercontent.com/10342151/131263693-f079ec86-d855-40bb-8bcf-fee6e212ddca.png)

I had this kind of error before. I think it's because of python version. Will add a potential fix soon

bramstroker commented:
I have added a commit which maybe solves your error. Could you try again?

stefangries commented:
Sorry, still not working for me:

![image](https://user-images.githubusercontent.com/10342151/131265251-c5de4133-d3ac-4f70-82d0-7642efaaf77f.png)

stefangries commented:
Yes, it works now, thank you. 

However, with your version I always get a blank line between records in the CSV. This was already the case with your V1. In my Tasmota version I had fixed this with `newline=''` when opening the CSV (see [here](https://github.com/bramstroker/homeassistant-powercalc/blob/a0eec65be1dd046eded23cf07083c282e7f4f681/utils/measure/measure_tasmota.py#L90)). But I don't know why this happens apparently only with me. 

Furthermore you take so far only one value. Therefore I receive with your version also still more noise than with mine. Maybe it makes sense either to take several values and calculate the average (like with my version) or later not only to interpolate, but also to smooth and remove outliers.

bramstroker commented:
> Yes, it works now, thank you.
> 
> However, with your version I always get a blank line between records in the CSV. This was already the case with your V1. In my Tasmota version I had fixed this with `newline=''` when opening the CSV (see [here](https://github.com/bramstroker/homeassistant-powercalc/blob/a0eec65be1dd046eded23cf07083c282e7f4f681/utils/measure/measure_tasmota.py#L90)). But I don't know why this happens apparently only with me.

Ok thanks for noticing. Will incoorporate this in the code.
> 
> Furthermore you take so far only one value. Therefore I receive with your version also still more noise than with mine. Maybe it makes sense either to take several values and calculate the average (like with my version) or later not only to interpolate, but also to smooth and remove outliers.

It's already in the task list.

bramstroker commented:
@stefangries Taking multiple samples is also implemented. You can set `SAMPLE_COUNT=3` in .env file.
Also added the requirements.txt
Only thing left is implement a mechanism to check if measurements are recent enough.

bramstroker commented:
First version of new script is finished. @CM000n @stefangries @smonesi @nepozs @OzGav, are any of you able to test the new tooling and give me feedback before I release this to public?
https://github.com/bramstroker/homeassistant-powercalc/tree/feature/measure_v2 here you can download or checkout the code. In `utils/measure_v2` is a README.md which should you get up and running.

smonesi commented:
I tried the kasa implementation, it doesn't work (it keeps looping on sample 0 until it stops).

As usual, I'm not a python expert, but this does seem to fix (since it looked like the "default" time.time() was not accepted):

        return PowerMeasurementResult(float(power),time.time())

I appreciate the new feature of taking more samples, as I feel like this is the reason why most measurements didn't seem right (eg. the standby value sometimes was higher than a low brightness value). However, it still looks like either the hue bulb (LCT015) is doing something strange or my HS110 plug is highly unreliable, because for a long period of time the average measurement of the standby value was 0.49W (which seems correct) but a few hours later it was more like 0.34W:

<img width="1564" alt="Screenshot 2021-08-30 at 23 26 58" src="https://user-images.githubusercontent.com/63957397/131408099-3dfec270-f653-4834-8eee-06abcbbf453c.png">

<img width="1571" alt="Screenshot 2021-08-30 at 23 27 22" src="https://user-images.githubusercontent.com/63957397/131408118-42092eca-916f-4530-812c-b1d7227bccd7.png">

In both cases, measurements can be as low as 0.15W or as high as 0.54W, it looks like at least 10 samples are needed to get a more sensible value.

Has anybody else noticed this behaviour? Should I stop using the HS110 for measurement since it is simply too unreliable?

bramstroker commented:
@smonesi Yes I think in `powermeter.py` default value is set to current time when script is first loaded into memory. Will see if I can fix this. Passing in `time.time()` from Kasa meter is perfectly fine.

I think this accuracy issue is mostly existent in the low power regions <0.5. most powerplugs have problems measuring this accurately. Shelly plug can't even measure below 0.4 watt.
@nepozs was testing the same plug afaik. Did you experience the same issues with HS110?

stefangries commented:
> Has anybody else noticed this behaviour? Should I stop using the HS110 for measurement since it is simply too unreliable?

I have the same problem with my tasmotas. So it can even happen that if you change a light regarding brightness from 1 to 2, that the power consumption according to the measurement decreases a bit instead of increasing. Apparently most WiFi plugs have this inaccuracy. If we can identify a plug that is more accurate, we should recommend it.

OzGav commented:
> Apparently most WiFi plugs have this inaccuracy.
Actually I think this is normal. All of the bulbs I have measured start at value X and then go down a little as the brightness increases then will jump up at some value then go down a little as the brightness increases further and it keeps doing that through the whole brightness range.

bramstroker commented:
@stefangries Could you work out a way to know the update time of the power sensor? Is this somehow exposed by tasmota endpoint? I did some google but couldn't found any references to `/?m`, not sure what kind of API this is.
If this is not possible `time.time()` is also fine, but than the feature to check for outdated measurements will not work.

<img width="817" alt="Screenshot 2021-08-31 at 09 09 07" src="https://user-images.githubusercontent.com/2345875/131458740-a18e8285-8cfa-461a-b10e-a73916dcd8f7.png">

bramstroker commented:
> Apparently most WiFi plugs have this inaccuracy. If we can identify a plug that is more accurate, we should recommend it.

Yes I also think this is normal behaviour. This is de way dimming works on LED lights, this is not 100% linear. @nepozs also created some graphs for it.

smonesi commented:
Latest commit fixes kasa 👍 

I'm currently trying to do some manual measurement with a Fibaro FGWPF-102 Z-Wave Plug, trying to understand if it is more precise (or at least more "stable") than the HS110, I'll post some graph here.

Is anybody able to write a powermeter that connects to the Z-Wave JS server? I did try but got stuck immediately since I couldn't find any meaningful example.

bramstroker commented:
> Latest commit fixes kasa 👍
> 
Great

> Is anybody able to write a powermeter that connects to the Z-Wave JS server? I did try but got stuck immediately since I couldn't find any meaningful example.
Did you see this library: https://github.com/home-assistant-libs/zwave-js-server-python? You'll have to make sure to map the websocket port from the container to the host so you can access the websocket from outside home assistant.

<img width="1203" alt="Screenshot 2021-08-31 at 18 55 01" src="https://user-images.githubusercontent.com/2345875/131544457-c8faab6a-beba-4350-9616-fdd406cabecf.png">

smonesi commented:
> Did you see this library: https://github.com/home-assistant-libs/zwave-js-server-python? You'll have to make sure to map the websocket port from the container to the host so you can access the websocket from outside home assistant.

Yes... but I still have no clue on how to interact with my Z-Wave Plug (ie. how to write actual code to ask for the power measure).

The result of my manual test shows that the Fibaro FGWPF-102 Z-Wave Plug only provides 1 decimal value but is much more stable compared to the HS110. I can provide some graph that shows this, if anybody is interested.

bramstroker commented:
> Yes... but I still have no clue on how to interact with my Z-Wave Plug (ie. how to write actual code to ask for the power measure).
> 
I also don't have an exact clue yet. We will need to connect to the websocket and than listen for node value messages somehow. I think we need to connect to the client and than call `listen`, https://github.com/home-assistant-libs/zwave-js-server-python/blob/master/zwave_js_server/client.py#L218. This will listen for messages and than construct event objects which should be handled somehow. Will see if I can do some coding next week.
